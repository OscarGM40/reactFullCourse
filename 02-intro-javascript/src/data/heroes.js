//exportaciones multiples.Se declaran constantes
//export const heroes = [
//si es una sola podemos usar por defecto.No se usa const

/* eslint import/no-anonymous-default-export: [2, {"allowArray": true}] */
//export default [
    const heroes =[
   {
       id: 1,
       name: 'Batman',
       owner: 'DC'
   },
   {
       id: 2,
       name: 'Spiderman',
       owner: 'Marvel'
   },
   {
       id: 3,
       name: 'Superman',
       owner: 'DC'
   },
   {
       id: 4,
       name: 'Flash',
       owner: 'DC'
   },
   {
       id: 5,
       name: 'Wolverine',
       owner: 'Marvel'
   },
];

const owners = ['DC','Marvel'];

//export default heroes;

export {
    owners,
    heroes as default,
}