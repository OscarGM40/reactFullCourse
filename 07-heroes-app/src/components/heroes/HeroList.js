import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

//publisher va a ser la editorial sol hay dos marvel y DC
export const HeroList = ({publisher}) => {
//vamos a usar useMemo para mayor eficiencia,pues si el publisher no cambia esta funcion no deberia ejecutarse y lo harÃ¡ si cambian las props,el state,etc...
  const heroesMemorized = useMemo(() => 
  getHeroesByPublisher(publisher),[publisher]);
  
  
  //const heroes = getHeroesByPublisher(publisher);

  return (
    <div className="card-columns animate__animated animate__fadeInDown ">
      {
        heroesMemorized.map( hero => (
          <HeroCard key={ hero.id } {...hero}> 
        </HeroCard>
        ))
      }
    </div>
  )
}
