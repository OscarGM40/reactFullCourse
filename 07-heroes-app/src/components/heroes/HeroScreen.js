import React, { useMemo } from "react";
import { Redirect, useParams } from "react-router-dom";
import { getHeroById } from "../../selectors/getHeroById";

export const HeroScreen = ({ history }) => {
  
  const { heroeId } = useParams();
  
  const heroMemorized = useMemo(() => getHeroById(heroeId), [heroeId]);
  //const hero = getHeroById(heroeId);
  

  if (!heroMemorized) {
    return <Redirect to="/" />;
  }

  const handleReturn = () => {
    if (history.length <= 2) (heroeId.includes('marvel')) ? history.push("/marvel") : history.push("/dc")
    else  history.goBack();
  };

  // console.log(heroMemorized)
  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
  } = heroMemorized;

  return (
    <div className="row mt-5 ">
      <div className="col-4">
        <img
          src={`./assets/heroes/heroes/${heroeId}.jpg`}
          alt={superhero}
          className="img-thumbnail animate__animated animate__bounceInRight"
        />
      </div>
      <div className="col-8">
        <h3> {superhero} </h3>

        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego: </b>
            {alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher: </b>
            {publisher}
          </li>
          <li className="list-group-item">
            <b>First Appearance: </b> {first_appearance}
          </li>
        </ul>
        <h5>Characters </h5>
        <p>{characters}</p>

        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};
