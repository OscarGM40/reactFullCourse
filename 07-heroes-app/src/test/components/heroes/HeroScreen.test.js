import { mount, shallow } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import { HeroScreen } from "../../../components/heroes/HeroScreen"


describe('Pruebas en HeroScreen', () => {

const historyMock = {
   length: 10,
   goBack: jest.fn(),
   push: jest.fn(),
}


test('Debe de mostrar el componente <Redirect> si no hay argumentos en el URL', () => {
   //no va a existir el heroe asi que redireccionara
   const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
         <HeroScreen />
      </MemoryRouter>
   )
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('Redirect').exists()).toBe(true);
      
   })

   test('Debe de mostrar un hero si el parametro existe y se encuentra', () => {
      //necesitamos usar useParams
      const wrapper = mount(

         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route path="/hero/:heroeId"  component={ HeroScreen }/>
         </MemoryRouter>
      )
      //con que existiera la clase row ya estriamos viendo el componente HeroCard!
         expect(wrapper.find('.col-4').exists()).toBe(true);

   })

   test('Debe de regresar a la pantalla anterior con PUSH', () => {
      //primero testearemos si el historial es menor a 2
      const historyMock = {
         length: 1,
         goBack: jest.fn(),
         push: jest.fn(),
      }
      //De alguna manera debemos mandarle el Mock
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route 
            path="/hero/:heroeId"  
            component={ (props) => <HeroScreen history={ historyMock }/> }/>
         </MemoryRouter>
      )
         wrapper.find('button').prop('onClick')();

         expect( historyMock.push ).toHaveBeenCalledWith('/marvel')
         expect( historyMock.goBack ).not.toHaveBeenCalled();
   })

   test('Debe de llamar a history.GOBACK,es decir,regresarÃ¡ a la pantalla anterior', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route path="/hero/:heroeId"
            component={ (props) => <HeroScreen history={ historyMock }/>} />
         </MemoryRouter>
      )
      wrapper.find('button').prop('onClick')();

      expect( historyMock.push).toHaveBeenCalledTimes(0);
      expect( historyMock.push ).not.toHaveBeenCalled();
      expect( historyMock.goBack).toHaveBeenCalled();
   })

   test('Debe de llamar el Redirect si el Hero no existe', () => {
        
      const wrapper = mount(
         <MemoryRouter initialEntries={['/hero/marvel-spider454354352']}>
            <Route path="/hero/:heroeId"
            component={ (props) => <HeroScreen history={ historyMock }/>} />
         </MemoryRouter>)

         console.log(wrapper.find('Redirect').exists());
         expect(wrapper.find('Redirect').exists()).toBeFalsy();
         expect(wrapper.text()).toBe('');
      
   })
   
})
