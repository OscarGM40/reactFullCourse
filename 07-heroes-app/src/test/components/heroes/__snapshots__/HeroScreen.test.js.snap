// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HeroScreen Debe de mostrar el componente <Redirect> si no hay argumentos en el URL 1`] = `""`;
